### ISSUE TITLE
Too wide parameter window 

### ISSUE THREAD

**Issue by ceraolo** [OpenModelica/OpenModelica#11721]
### Description
In some cases the parameter window is too large.
I'm not sure whether a ticket on this already exists. I searched and did not find anything.

### Steps to Reproduce
Open combi in the following model:
```
model CombiTest
parameter Real Cgen=1;
parameter Real Cload=1;
Modelica.Blocks.Tables.CombiTable1Ds combi(smoothness = Modelica.Blocks.Types.Smoothness.ContinuousDerivative, table = [0, 1; 1, 1]) annotation(
      Placement(transformation(origin = {-8, 0}, extent = {{-10, -10}, {10, 10}})));
Modelica.Blocks.Sources.RealExpression combiIn(y = time)  annotation(
      Placement(transformation(origin = {-52, 0}, extent = {{-10, -10}, {10, 10}})));
  equation
connect(combiIn.y, combi.u) annotation(
      Line(points = {{-40, 0}, {-20, 0}}, color = {0, 0, 127}));
  annotation(
      Diagram(coordinateSystem(extent = {{-80, 20}, {20, -20}})),
  version = "",
  uses);
end CombiTest;

```

The following parameter window is opened (and cannot be horizontally reduced). It is definitely too wide.
![image](https://github.com/OpenModelica/OpenModelica/assets/7373030/5ddb9a26-bafb-461b-8f08-0bbb29495cfb)

### Expected Behavior
The window width should be adequate to the content, and it should  be always possible to shrink it, up to a reasonable (very low) minimum.

### Version and OS
OM 1.23.0-dev 193

**Comment by abuntrock** [comment]
Ticket #10395 is probably related to this

**Comment by abuntrock** [comment]
Is there a timeline regarding a fix for this issue?

**Comment by casella** [comment]
No, but we can prioritize it if you need.

**Comment by abuntrock** [comment]
After the performance tickets, the possibility of a conversion #8806, and an editor for array parameters #11048, this would be the next issue in our list of importance.

**Comment by casella** [comment]
OK, I updated https://github.com/OpenModelica/OpenModelica/projects/2 accordingly

**Comment by abuntrock** [comment]
Is there any work planned regarding this issue for OMEdit 1.26?

**Comment by casella** [comment]
I will prioritize this for 1.26.0

**Comment by casella** [comment]
@adeas31 this should really be done for 1.26.0.

**Comment by casella** [comment]
Requirements
- sizing should show all the textbox contents
- comments shown in full only with tooltips (on multiple lines), otherwise they get cropped

**Comment by adeas31** [comment]
> Requirements
> 
> * sizing should show all the textbox contents
> * comments shown in full only with tooltips (on multiple lines), otherwise they get cropped

PR #14723 does the following changes,

 - The comment of parameters on the right end are now shown with ellipses and full comment is available via tooltip.
 - The fields are not stretch to show full values.
 - The dropdown items are also shown with ellipses and full text can be seen in the tooltip.

Try the latest nightly tomorrow and let me know if you need more changes.

**Comment by ceraolo** [comment]
> Try the latest nightly tomorrow and let me know if you need more changes.

When reducing width I think OMEdit should first eliminate unused spaces (e.g. blank spaces in edit fields), then compact texts. 
I checked the recent changes with this ticket's example and I saw the expected behaviour only partially.

As you can see In the screencast below, when I reduce space, text gets compacted when still large chunks of white space still exist in edit fields. I cannot see any good reason to have this. Do I miss something?

https://github.com/user-attachments/assets/a6e95aad-3526-4cb6-b041-53a8b504836f



Moreover, there are cases (I could not understand when)  in which the sizing is worse than the screencast above. See the one below:

https://github.com/user-attachments/assets/c8585ae0-37f6-4806-bfcf-345d6accca17

**Comment by adeas31** [comment]
The actual problem are comboboxes and not the edit fields with whitespace. They are just stretched out because of the dropdown items of `smoothness`.  I set the items width to maximum of 500 pixels, items larger than that will be shown with ellipses.

If I set the fixed width for comboboxes than they will not resize when you resize the dialog and it is very bad design to set a fixed width.

**Comment by ceraolo** [comment]
>I set the items width to maximum of 500 pixels, items larger than that will be shown with ellipses.

The behaviour we see on my first screencast shows that we have more a _minimum_ limit than a _maximum_: when I resize (while reducing) the dialog box, combo boxes initially reduce their width then do not do this anymore: they have reached their minimum.

In general I see a major  drawback in setting limits in pixels:  a limit in pixel may be too small or large depending on the screen resolution: a limit of 500 pixels would have a totally different visual impact on a 1k or 4k screen. 
Moreover, your static minimum causes those blank parts to spoil some of the precious desktop area.

I propose this: set the combo boxes user size to the _maximum_ between
- the actual size of the displayed string (with the current font at the current screen resolution: I remember this can be easily evaluated in Qt)
- a fixed value corresponding to the size (with the current font at the current screen resolution) of a test string, that could be, for instance, "XXX XXX"

In this way, the combo box size will always have the width needed to display the actual text, with the exception of very small content, requiring less amount of horizontal space than the test string.

This strategy would need recalculation of the combo width every time a new item is chosen, but I suppose it is not a huge programming burden.


Probably you can also set a limit in combo box width that must not be overcome, with padding with ellipses in case of long combo box text. Not sure what to choose for this limit. Maybe half of the screen width? A test implementation could help.

>If I set the fixed width for comboboxes than they will not resize when you resize the dialog and it is very bad design to set a fixed width.

I never thought of using a fixes width for comboboxes

**Comment by casella** [comment]
> I propose this: set the combo boxes user size to the _maximum_ between
> 
>     * the actual size of the displayed string (with the current font at the current screen resolution: I remember this can be easily evaluated in Qt)
> 
>     * a fixed value corresponding to the size (with the current font at the current screen resolution) of a test string, that could be, for instance, "XXX XXX"

@adeas31 this seems to me a good proposal, if it's not too complicated to implement. The text string could be something like "MMMMM MMMMMM MMMMM MMMMMM MMMM". This is slightly less than half the string width on my laptop screen, but will be much less on a 4K screen, which looks OK.

**Comment by ceraolo** [comment]
> @adeas31 this seems to me a good proposal, if it's not too complicated to implement. The text string could be something like "MMMMM MMMMMM MMMMM MMMMMM MMMM". This is slightly less than half the string width on my laptop screen, but will be much less on a 4K screen, which looks OK.

If you propose such a long string I failed to explain my proposal. 
My proposal wants the combo box to avoid having white spaces if space is needed for the descriptive string. Now the descriptive string is cut (with ellipses), while the combo box has thite space, which I dislike.

I added the short "test string" only to avoid seeing a very short combo box in rare cases, but it is not mandatory. We can substitute "XXX XXX" with "XXX" or even "", not with long strings!

**Comment by adeas31** [comment]
> The behaviour we see on my first screencast shows that we have more a minimum limit than a maximum: when I resize (while reducing) the dialog box, combo boxes initially reduce their width then do not do this anymore: they have reached their minimum.

What I mean is that the items only show the text that can fit inside 500 pixels otherwise the text will be shown with ellipses so in a way it is minimum width restriction.

> the actual size of the displayed string (with the current font at the current screen resolution: I remember this can be easily evaluated in Qt)

Combobox also take in account the size of its items. And it sets the width of the combobox to the maximum item size. In your first video the size of combobox is 500 becasue of its items. Note that we display the name + comment in the items but when you select the item we only use name. One solution is to only display names and show name + comment in tooltip of items.

> In this way, the combo box size will always have the width needed to display the actual text, with the exception of very small content, requiring less amount of horizontal space than the test string.

> This strategy would need recalculation of the combo width every time a new item is chosen, but I suppose it is not a huge programming burden.

That is possible but will be a very bad design idea since that will cause a repaint for the whole dialog and also you will get a jumping dialog which is changing its width.

**Comment by ceraolo** [comment]
> 
> What I mean is that the items only show the text that can fit inside 500 pixels otherwise the text will be shown with ellipses so in a way it is minimum width restriction.

These 500 pixels, in my opinion, make things too rigid. In general, I think that using fixed numbers of pixels when choosing the graphical aspect of dialog boxes always creates visual shortcomings, because its effect is dependent on screen resolution and font size.
(I understand that in this case this limit determines the maximum width of the combo box width, and causes the comment after the name to be shortened with ellipses.)

> 
>  Note that we display the name + comment in the items but when you select the item we only use name. 

Ok, I did not realise this earlier.

>One solution is to only display names and show name + comment in tooltip of items.

good idea!

> 
> That is possible but will be a very bad design idea since that will cause a repaint for the whole dialog and also you will get a jumping dialog which is changing its width.

I knew that this would cause repainting the dialog box. I presumed this should not be bad for the user experience (but I expected confirmation in evaluating this user experience on a prototype implementation). 
Since you are sure this is bad, I propose now to use as a minimum size not the size of the actually displayed item, but, instead, the size of largest of all the combo items possible (size in pixels measured at the current resolution and with  the current font).

So, the situation should be rather clear:
1. only display names and show name + comment in tooltip of items.
2. let the combo box to collapse up to the size of the largest name in the combo list.

Possible addition:
3. to avoid strange behaviour in case of very, very long combo names (which are indeed very unlikely), a total maximum for the combo boxes can be added, e.g. one half of the available desktop width  in pixels (which is dependent on the screen resolution). Not sure this is actuallly needed: limiting the displaied value to only names, (not names + comments) should ensure that combo boxes nearly never to become too wide. If I were you, I would make some tests to see whether this third point is necessary or not.

**Comment by casella** [comment]
> One solution is to only display names and show name + comment in tooltip of items.

The only (potential) problem I see of this solution is that sometimes the name is not really that informative, e.g.:
- Modelica.Blocks.Types.Smoothness.Monotone.ContinuousDerivative1
- Modelica.Blocks.Types.Smoothness.Monotone.ContinuousDerivative2
which should I pick? Only reading the comments allows one to choose:
- Fritsch-Butland spline interpolation (such that the monotonicity is preserved and the first derivative is continuous)
- Steffen spline interpolation of table points (such that the monotonicity is preserved and the first derivative is continuous)

So, it's important that the tootip shows up immediately, so you can quickly see the different comments by moving around the mouse. Other than that, it seems a good solution to me.

> 3. to avoid strange behaviour in case of very, very long combo names (which are indeed very unlikely)

Yeah, I wouldn't be too worried about this. We can implement the solution stated above and see what happens in practical cases.

**Comment by abuntrock** [comment]
I do not think that I can add anything valuable to the discussion regarding width in general. @adeas31 and @ceraolo you seem much more competent than me to decide what works and what not with respect to resolution, fixed pixel withs etc.. As I wrote in #10395 the overall behavior now looks much more pleasing for me.

I just want to stress that for us at Bosch Rexroth the comment is nearly always more important than the name. We use replaceable objects a lot to switch between different parameterizations of one hydraulic pump for example. Since the comment is not bound to Modelica naming rules we can use the actual product ordering codes as comments including for example slashes (see picture below). 

<img width="983" height="94" alt="Image" src="https://github.com/user-attachments/assets/c1877fb0-6eb9-4ea9-9e92-037cb462eaf9" />

This is why we add the Dymola specific annotation `__Dymola_includeChoice=false` to basically all of our replaceable declarations.

1. Would it be possible to add an option if the name or the comment should be shown?
2. In my opinion it is still unnecessary to add the whole redeclare syntax to all dropdown menu items (like in the picture above). For users, which are not se deep into Modelica, like the ones from our BRSL library, this adds no value. Just showing the name or the comment (we would prefer comment) of the different choices would be sufficient. I know that this was actively decided so in https://github.com/OpenModelica/OpenModelica/pull/13264 but I wanted to make one last argumentative effort because it would probably reduce the width needed by those replaceable boxes.

**Comment by ceraolo** [comment]
> 1. Would it be possible to add an option if the name or the comment should be shown?

No proposal asked to drop the comment. It was proposed just to let it visible as a tooltip (not as both text and tooltip). Tooltips are very well visible in yellow background.

**Comment by adeas31** [comment]
This is up for review again @casella @ceraolo @abuntrock 

With PR #14774 the drop down items only show minimal text i.e., name of enumerations/choices. The detailed information like comment is shown in tooltip.

The dropdown minimum size is set to the size of largest item. Other than that the dialog doesn't force any size restrictions and rely on qt's default layout.

**Comment by casella** [comment]
> No proposal asked to drop the comment. It was proposed just to let it visible as a tooltip (not as both text and tooltip). Tooltips are very well visible in yellow background.

Yes, the problem with comments is that sometimes they can be really long. Fitting them in a single line leads to these ugly wide parameter windows, while cutting the with ellipses is really nasty. Tooltip text is nice because it can be fit on multiple lines.

**Comment by abuntrock** [comment]
Understood. Our usecase there is probably one of many. I like the tooltip appearance in general.
So we can with that. The current situation is definitely way better than having those wide parameter windows.

**Comment by ceraolo** [comment]
@adeas31 I have checked the latest nightly build and I like your implementation.

There still exists the issue that I showed in my second video, that I replicate again in the following one:

https://github.com/user-attachments/assets/468377a1-a0a8-466c-acff-6d7515a1d709

You see, in Ubuntu, I cannot enlarge the window the size I want.

**Comment by adeas31** [comment]
That's so weird. I don't know why this is happening. I can't reproduce this on Windows.

**Comment by perost** [comment]
I tried it on Linux with Cinnamon/X11, and can enlarge the window as much as I want. So it might be an issue with either a specific desktop environment (Gnome?) or maybe Wayland.

**Comment by ceraolo** [comment]
I made tests on Ubuntu 24.04.3 LTS (in an Oracle virtual box 7.0),

So:
1. the issue might be related to the specific desktop environment, as @perost says 
2. it might even be related to my Virtual box virtualisation software
3. might be real (whatever this might mean).

I use Ubuntu only to make fast tests of improvements on OM (BTW, this because upgrading there takes just 1-2 min).
For the rest of my many OM-related activities  I use Win or Mac (via docker).

**So, for me, this ticket can be closed.** It could be reopened just in case some real Ubuntu user experiences mulfunctioning.

But I leave the final decision to the programmers.
Thanks a lot, @adeas31 !

**Comment by casella** [comment]
If @ceraolo and @abuntrock are happy, who am I to disagree? 😄 

Seriously, feel free to propose further improvements in future tickets.
